src/librustc_mir/build/expr/into.rs

```diff
@@ -163,18 +163,18 @@ impl<'a,'tcx> Builder<'a,'tcx> {
                         body_block = loop_block;
                     }
 
-                    // execute the body, branching back to the test
-                    // We write body’s “return value” into the destination of loop. This is fine,
-                    // because:
-                    //
-                    // * In Rust both loop expression and its body are required to have `()`
-                    //   as the “return value”;
-                    // * The destination will be considered uninitialised (given it was
-                    //   uninitialised before the loop) during the first iteration, thus
-                    //   disallowing its use inside the body. Alternatively, if it was already
-                    //   initialised, the `destination` can only possibly have a value of `()`,
-                    //   therefore, “mutating” the destination during iteration is fine.
-                    let body_block_end = unpack!(this.into(destination, body_block, body));
+                    // The “return” value of the loop body must always be an unit, but we cannot
+                    // reuse that as a “return” value of the whole loop expressions, because some
+                    // loops are diverging (e.g. `loop {}`). Thus, we introduce a unit temporary as
+                    // the destination for the loop body and assign the loop’s own “return” value
+                    // immediately after the iteration is finished.
+                    let tmp = this.get_unit_temp();
+                    // Execute the body, branching back to the test.
+                    let body_block_end = unpack!(this.into(&tmp, body_block, body));
+                    // Even if we assign unconditionally, if the loop turns out to be diverging,
+                    // the loop’s “return” value will not appear in the MIR (removed along with the
+                    // exit_block by cfg simplification passes later).
+                    this.cfg.push_assign_unit(exit_block, expr_span, destination);
```

Comment:
I'm actually a little bit surprised by the behavior of the type checker here. It seems like `loop { }` has type unit -- *but*, if you never `break` out, we consider the loop non-terminating, and hence we allow it to be used in any context. This is probably not how I would want it to work in an ideal world, but it may not be worth trying to change anything here. :) 

Therefore, to match the compiler as today, I suspect it should work like this:

- if you break out of the loop, then we will `push_assign_unit`
- if you do not, we will not push any assignment at all

Of course we don't know whether you break out of the loop in advance. We could either analyze this, or maybe come back after the fact and check.